ddl/ddl_api.go

```diff
@@ -3718,6 +3718,208 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
+	// 2 Create the new partitions
+	//   Including placement etc.
+	// 3 create a DDL job for it (including implementing the ddl_worker counter part...
+
+	// Various checks
+	switch pi.Type {
+	// Only supporting RANGE/LIST
+	case model.PartitionTypeRange:
+		// TODO: Check that the new partitions are in one range, not overlapping the kept ones.
+	case model.PartitionTypeList:
+	default:
+		return errors.Trace(dbterror.ErrUnsupportedReorganizePartition)
+	}
+	firstPartIdx, lastPartIdx, idMap, err := getReplacedPartitionIDs(spec.PartitionNames, pi)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if pi.Type == model.PartitionTypeRange {
+		if len(idMap) != (lastPartIdx - firstPartIdx + 1) {
+			// Not continues range
+			// TODO: Better error message?
+			return errors.Trace(dbterror.ErrUnsupportedReorganizePartition)
+		}
+		// TODO: Check that the new partitions are not overlapping the kept ones.
+		// I.e. lastPartIdx partition needs either to be the last partition
+		//   OK to increase/decrease range, will error out if any row is out of the new range)
+		// or partition definition range, must be the same as the last partition it replaces
+	}
+
+	partInfo, err := BuildAddedPartitionInfo(ctx, meta, spec)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err = d.assignPartitionIDs(partInfo.Definitions); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err = checkReorgPartitionDefs(ctx, meta, partInfo, firstPartIdx, lastPartIdx, idMap); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err = handlePartitionPlacement(ctx, partInfo); err != nil {
+		return errors.Trace(err)
+	}
+
+	job := &model.Job{
+		SchemaID:   schema.ID,
+		TableID:    meta.ID,
+		SchemaName: schema.Name.L,
+		TableName:  t.Meta().Name.L,
+		Type:       model.ActionReorganizePartition,
+		BinlogInfo: &model.HistoryInfo{},
+		Args:       []interface{}{spec.PartitionNames, partInfo},
+		ReorgMeta: &model.DDLReorgMeta{
+			// We should not depend on SQL Mode (TODO: check vs MySQL for zero date etc.)
+			SQLMode:       mysql.ModeNone,
+			Warnings:      make(map[errors.ErrorID]*terror.Error),
+			WarningsCount: make(map[errors.ErrorID]int64),
+			// We should not evaluate any new values? (TODO: Check with virtual/generated columns with indexes?)
+			Location: &model.TimeZoneLocation{Name: time.UTC.String(), Offset: 0},
+		},
+	}
+
+	err = d.DoDDLJob(ctx, job)
+	if err == nil {
+		d.preSplitAndScatter(ctx, meta, partInfo)
+	}
+	err = d.callHookOnChanged(job, err)
+	return errors.Trace(err)
+}
+
+func checkReorgPartitionDefs(ctx sessionctx.Context, tblInfo *model.TableInfo, partInfo *model.PartitionInfo, firstPartIdx, lastPartIdx int, idMap map[int]interface{}) error {
+	// partInfo contains only the new added partition, we have to combine it with the
+	// old partitions to check all partitions is strictly increasing.
+	pi := tblInfo.Partition
+	clonedMeta := tblInfo.Clone()
```

Comment:
IMHO the semantic of `TableInfo.Clone()` should have `TableInfo.Partition` cloned, but it doesn't so :( maybe we could add it, so you don't have to clone the `Partition` here?

Reply:
Fixed, I added `Partiton` handling into `TableInfo.Clone()`.